### ISSUE TITLE
Required module not found

### ISSUE THREAD

**Issue by highdali** [brunocodutra/flow-bin-loader#3]
# error
```
src/containers/index.js:3
  3: import App from './App.js'
                     ^^^^^^^^^^ ./App.js. Required module not found
```

# Directory Structure

```
 .
+--_ .flowconfig
+--_ src
|    +-----_ components
|    |     +-------- Header.js
|    |
|    +-----_ containers
|    |     +-------- App.js
|    |     +-------- index.js
|    |
|    +-----index.js
|
+--_ .flowconfig
+--_ .eslintrc
+
+--_ webpack.config.js
```

# .flowconfig
```js
[ignore]

[include]

[libs]

[options]
module.name_mapper='^\(.*\)$' -> '<PROJECT_ROOT>/src/\1'
module.file_ext=.js
module.file_ext=.json
module.file_ext=.jsx
module.file_ext=.css
module.file_ext=.scss
```

# .eslintrc
```js
{
    "env": {
        "browser": true,
        "commonjs": true,
        "es6": true,
        "node": true
    },
    "parser": "babel-eslint",
    "parserOptions": {
        "ecmaFeatures": {
            "jsx": true
        },
        "sourceType": "module"
    },
    "extends": [
        "plugin:flowtype/recommended",
        "eslint:recommended",
        "plugin:react/recommended"
    ],
    "plugins": [
        "flowtype",
        "babel",
        "react"
    ],
    "rules": {
        "no-console": "off",
        "no-unused-vars": "warn"
    }
}
```

# src/index.js
```js

/* @flow */
import React from 'react'
import { render } from 'react-dom'

import AppContainer from './containers'

render(
  <div>
    <AppContainer />
  </div>,
  document.getElementById('root')
);
```

# src/containers/index.js
```js
/* @flow */

import App from './App.js'

const AppContainer = App

export default AppContainer
```

# src/containers/App.js
```js
/* @flow */

import React, { Component } from 'react'

import Header from '../components/Header'

class App extends Component {
  render() {
    return (
      <Header />
    );
  }
}

export default App
```

# webpack.config.js

```js
var webpack = require('webpack')

var path = require('path');

module.exports = {
  entry: ['babel-polyfill', `${__dirname}/src/index.js`, `${__dirname}/src/stylesheets/main.scss`],
  output: {
    path: `${__dirname}/public`,
    filename: 'bundle.js'
  },
  module: {
    rules: [
      {
        enforce: 'pre',
        test: /\.jsx?$/,
        use: ['flow-bin-loader', 'eslint-loader'],
        exclude: /node_modules/
      },
      {
        use: 'babel-loader',
        test: /\.jsx?$/,
        exclude: /node_modules/
      },
      {
        use: [
          'style-loader',
          'css-loader?importLoaders=1',
          'postcss-loader',
          'sass-loader'
        ],
        test: /\.scss$/
      },
      {
        use: [
          'style-loader',
          'css-loader?importLoaders=1',
          'postcss-loader'
        ],
        test: /\.css$/
      },
      {
        test: /\.(eot|svg|ttf|woff|woff2)$/,
        use: {
          loader: 'file-loader',
          options: {
            name: 'public/fonts/[name].[ext]'
          }
        }
      },
      {
        test: /\.(png|jpg)$/,
        use: {
          loader: 'url-loader',
          options: {
            limit: 10000
          }
        }
      }
    ]
  },
  plugins: [
    new webpack.HotModuleReplacementPlugin(),
    new webpack.LoaderOptionsPlugin({
      options: {
        postcss: [
          require('autoprefixer')
        ]
      }
    })
  ],
  devServer: {
    port: 4000,
    contentBase: `${__dirname}/public`,
    historyApiFallback: true,
    inline: true,
    hot: true
  }
}
```

**Comment by highdali** [comment]
In .flowconfig options, i think `module.name_mapper` should support multiple directory functionality for same pattern(`require(./foobar) in src/A)` and `require('./fooapp') in src)` ) etc.

**Comment by brunocodutra** [comment]
When you `import App from './App.js'`, flow looks within `src/` for `App.js`, because that is what  `module.name_mapper='^\(.*\)$' -> '<PROJECT_ROOT>/src/\1'` tells it to do, but it is not there, hence the error.

You should `import App from 'containers/App.js` instead.

>  i think `module.name_mapper` should support multiple directory functionality for same pattern

I believe you can achieve what you want by providing multiple `module.name_mapper=` rules. Flow should pick the first to match the pattern and apply the mapping, but I would advise against doing so as undesired name collisions tend to occur as the project grows. Instead I'd suggest that you stick to specifying the full path to files from `src/`.

**Comment by highdali** [comment]
`import App from 'containers/App.js'` does not work.
error shows
```
ERROR in ./src/containers/index.js
Module not found: Error: Can't resolve 'containers/App.js' in '/Users/highdali/dev/book-review-moa/src/containers'
 @ ./src/containers/index.js 7:11-39
 @ ./src/index.js
 @ multi (webpack)-dev-server/client?http://localhost:4000 webpack/hot/dev-server babel-polyfill ./src/index.js ./src/stylesheets/main.scss
```

Webpack module resolve engine found the 'containers/App.js' in /src/containers' not '/src'

**Comment by highdali** [comment]
For example project, I've updated example in my [repo](https://github.com/highdali/book-review-moa/tree/issue) issue branch.

**Comment by highdali** [comment]
similar trick:

In webpack config,

```js
resolve: {
    modules: [
      path.join(__dirname, "src"),
      path.join(__dirname, "src/components"),
      "node_modules"
    ]
  }
```

**Comment by brunocodutra** [comment]
I didn't have a chance to test your project locally yet, but does it help if you define `resolve.root` in `webpack.conf.js`?

```{.js}
resolve: {
    root: [path.join(__dirname, "src")],
    fallback: "node_modules",
    extensions: ['.js', '.json', '']
  },
```

**Comment by highdali** [comment]
I think it is also hack.

Imagine directory structure is like below...

```
 .
+------ webpack.config.js
+------ .flowconfig
+------ src/
|       +------ Hello.js
|       +-------index.js 
```

# src/index.js
```js
import React, { Component } from 'react'

/* Here, Why we import Hello component like this.......*/
import Hello from 'components/Hello'
/* 
I want to import Hello following directory structure
so I think it should be import Hello from './Hello' 
I want to do not import component against directory structure..........
*/

class Container extends Component {
  render() {
    return (
      <Hello />
    );
  }
}

export default Container
```

# src/Hello.js
```js
import React, { Component } from 'react';

class Hello extends Component {
  render() {
    return (
      <div>Hello</div>
    );
  }
}

export default Hello
```

**Comment by highdali** [comment]
`Instead I'd suggest that you stick to specifying the full path to files from src/.`
I think it's good to follow your advice. Thanks

**Comment by highdali** [comment]
It's not resolved.

Same #1

**Comment by brunocodutra** [comment]
Sorry for the delay.

I took a closer look at your project and in turns out my suggestions were correct, except that I missed the fact you are using webpack 2. 

The fix is to add the following to `webpack.config.js`

```
resolve: {
  modules: ["src", "node_modules"]
},
```

and use paths relative to `src/` to import modules, that is, `'components/Header'` instead of `../components/Header'` and `'containers/App'` instead of `'./App'`.

I've submitted a PR to your repository for reference, see highdali/book-review-moa#1

It works for me locally, let me know if it fixes it for you as well.
